### PR TITLE
A better take forward kernel for CPU

### PR DESCRIPTION
**Pull request by HyperZealot**
## Description ##
The same as title.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
~5x improvement on CPU with this kernel

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@HyperZealot Thanks for the efforts for this great improvements :)
Do you mind pointing out which model/test is tested with 5X  speedup?

**Comment by HyperZealot** [comment]
@pengzhao-intel Hello, I have encountered this problem on my own model so I don't want to share that directly, but I can provide a small script to measure this, would you like to see that?

**Comment by HyperZealot** [comment]
@pengzhao-intel 
script:
```Python
import mxnet as mx
import random
from mxnet.test_utils import rand_ndarray, check_speed

num_rows = 1000000
test_rows = 20000
num_cols = 512
shape = (num_rows, num_cols)
data = rand_ndarray(shape, stype='default')
print(data.dtype)
x = mx.sym.Variable('data')
idx = mx.sym.Variable('indices')
sym = mx.sym.take(a=x, indices=idx, axis=0)
trials = int(num_rows / test_rows)
total_time = 0.0
all_indices = [i for i in range(num_rows)]
random.shuffle(all_indices)
for i in range(trials):
    indices = mx.nd.array(all_indices[i * test_rows: (i + 1) * test_rows])
    total_time += check_speed(sym, location={'data': data, 'indices': indices}, ctx=mx.cpu(), N=100, typ="forward")
print(test_rows * num_cols * 4 * trials / total_time / 1024 / 1024 / 1024, "GB/s")
```
Before with mxnet-mkl release version: 2.0371343465121163 GB/s
After: 12.2486153514499 GB/s
~6x speedup with this new one

**Comment by pengzhao-intel** [comment]
@HyperZealot thanks for your info and script. We also have an optimized version of `take` in local and plan to upstream to the official repo. 
@rongzha1 will try your script and compare the performance of both implementations.
I think we can combine two solutions together  :)

File: src/operator/tensor/indexing_op.cc [hunk 1]
```diff
@@ -28,6 +28,28 @@
...
+struct TakeCPU {
+  // assume that idx have been flattened to a 1-D tensor (N,)
+  // assume that out_data and in_data have been flattened to 2-D tensors, (N, M) and (K, M)
+  // M is the number of columns of in_data and out_data
+  // K is the number of rows of in_data
+  // i is the index of out_data
+  template<typename DType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data, const DType* in_data,
+                                  const IType* idx, const size_t M, const int64_t K) {
+    int64_t j = static_cast<int64_t>(idx[i]);
```
Comment (rongzha1) [inline src/operator/tensor/indexing_op.cc]: Can you tell why  use idx[i]?  Origin is idx[i/M] 
Thanks

Comment (rongzha1) [reply src/operator/tensor/indexing_op.cc]: origin is 
int j = static_cast<int>(idx[i/M]);

Comment (rongzha1) [reply src/operator/tensor/indexing_op.cc]: OK I see. But what if N%M != 0,  the remaining N%M part ?

File: src/operator/tensor/indexing_op.cc [hunk 2]
```diff
@@ -48,6 +70,29 @@ bool CheckIndexOutOfBound(const DType* data_ptr, size_t data_size,
...
+  const TShape& oshape = output.shape_;
+
+  MSHADOW_TYPE_SWITCH(output.type_flag_, DType, {
+    MSHADOW_TYPE_SWITCH(data.type_flag_, IType, {
+      Tensor<cpu, 1, IType> idx = data.get_with_shape<cpu, 1, IType>(
+        Shape1(ishape.ProdShape(0, ishape.ndim())), s);
+      Tensor<cpu, 2, DType> wmat = weight.get<cpu, 2, DType>(s);
+      Tensor<cpu, 2, DType> out = output.get_with_shape<cpu, 2, DType>(
+        Shape2(oshape.ProdShape(0, oshape.ndim()-1), oshape[oshape.ndim()-1]), s);
+      Kernel<TakeCPU<true>, cpu>::Launch(s, oshape.Size() / wmat.shape_[1], out.dptr_, wmat.dptr_,
```
Comment (rongzha1) [inline src/operator/tensor/indexing_op.cc]: As above. What if   oshape.Size() % wmat.shape_[1]   != 0;   the remaining  oshape.Size() % wmat.shape_[1]  bit ?

File: src/operator/tensor/indexing_op.cc [hunk 3]
```diff
@@ -227,6 +272,74 @@ void TakeOpForwardCsrImpl<cpu>(const TakeParam& params,
...
+  const TShape& oshape = outputs[take_::kOut].shape_;
+
+  Stream<cpu> *s = ctx.get_stream<cpu>();
+  const int actual_axis = param.axis + ((param.axis < 0) ? arrshape.ndim() : 0);
+
+  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {  // output data type
+    MSHADOW_TYPE_SWITCH(inputs[1].type_flag_, IType, {  // index data type
+      if (actual_axis == 0) {
+        if (param.mode == take_::kClip) {
+          Kernel<TakeCPU<true>, cpu>::Launch(s, idxshape.Size(),
```
Comment (rongzha1) [inline src/operator/tensor/indexing_op.cc]: Same as above

File: src/operator/tensor/indexing_op.cc [hunk 4]
```diff
@@ -227,6 +272,74 @@ void TakeOpForwardCsrImpl<cpu>(const TakeParam& params,
...
+                                             outputs[take_::kOut].dptr<DType>(),
+                                             inputs[take_::kArr].dptr<DType>(),
+                                             inputs[take_::kIdx].dptr<IType>(),
+                                             oshape.Size()/idxshape.Size(), arrshape[0]);
+        } else {
+          Kernel<TakeCPU<false>, cpu>::Launch(s, idxshape.Size(),
```
Comment (rongzha1) [inline src/operator/tensor/indexing_op.cc]: same as above.  What if oshape.Size()%idxshape.Size() != 0?

**Comment by HyperZealot** [comment]
@rongzha1 Hi, all shapes here are inferred so they are guaranteed to satisfy divisibility requirements.

**Comment by ankkhedia** [comment]
@HyperZealot  Thanks for the awesome contribution!
@rongzha1 It seems all your comments have been addressed. Could you please review the PR again 

@mxnet-label-bot [pr-awaiting-review, operator]

**Comment by rongzha1** [comment]
> @rongzha1 Hi, all shapes here are inferred so they are guaranteed to satisfy divisibility requirements.

Thanks for your reminder
There's still one question.  When you use memcpy, can you add   if(size> N) ? 
Some model M is very small, such as wide_deep we are doing optimization (https://github.com/apache/incubator-mxnet/tree/master/example/sparse/wide_deep) M is 8. If you use memcpy to copy 8 float every time, overhead is much big.

I test it use your script using our wide deep parameters 
num_rows = 800
test_rows = 8
num_cols = 61400


  | M=8
-- | --
for | 131.7215735239637 GB/s
memcpy | 58.68885431408931 GB/s

File: src/operator/tensor/indexing_op.cc [hunk 5]
```diff
@@ -28,6 +28,28 @@
...
+                                  const IType* idx, const size_t M, const int64_t K) {
+    int64_t j = static_cast<int64_t>(idx[i]);
+    if (clip) {
+      if (j <= 0) j = 0;
+      else if (j >= K) j = K - 1;
+    } else {
+      j = j % K;
+      j += (j < 0) ? K : 0;
+    }
+    std::memcpy(out_data + i * M, in_data + j * M, M * sizeof(DType));
```
Comment (rongzha1) [inline src/operator/tensor/indexing_op.cc]: can you do some test to get a best some_size :
if( M > some_size) {
  std::memcpy(out_data + i * M, in_data + j * M, M * sizeof(DType))
} else {
      const int jM = j*M;
      const int iM = i*M;
      for (int n = 0; n < M; n++) {
        out_data[iM + n] = in_data[jM + n];
      }
}

Small M will have a big overhead

**Comment by HyperZealot** [comment]
@rongzha1 From your code M should be num_cols(61400), that is a pretty big number.

**Comment by rongzha1** [comment]
Thanks for your info.  I have asked admin about the server config: 
CPU: skylake 8180  (56 physical cores)
Memory bandwidth :
255GB/s = 2.66*12*8(2666MHz DDR4) 
and run STREAM:  (omp use all 56 core)
-------------------------------------------------------------
Function    Best Rate MB/s  Avg time     Min time     Max time
Copy:          129854.6     0.001265     0.001232     0.001322
Scale:         121861.0     0.001392     0.001313     0.001653
Add:           115944.8     0.002097     0.002070     0.002171
Triad:         126780.0     0.001911     0.001893     0.001948

maybe some config error, when I set OMP_NUM_THREADS=28 (half its total cores), performance even better.

**Comment by HyperZealot** [comment]
@rongzha1 TBH that's also what I got on my side. my CPU has 4 physical cores, so when I'm using OMP_NUM_THREADS=4 I get the best performance (~30GB/s).

**Comment by HyperZealot** [comment]
@rongzha1 BTW I don't think your machine is a representative example for most users'...

**Comment by pengzhao-intel** [comment]
@HyperZealot FYI, This is a different scenario for the enterprise users who will use the high bin machines (servers, like SKX 6148, 8180) with multiple cores (20+) per socket and high bandwidth memory for their services, such as AWS cloud. So, we need to make the change more general to cover various cases.

Thanks for the improvements again :)

**Comment by HyperZealot** [comment]
I don't think the workloads provided by @rongzha1 is suitable for determining the memory bandwidth, based on the following arguments:
0. For each input, the effective working set size is test_rows*num_cols*4 bytes. Here's the working set size for each input:
(1M, 20k, 512) = ~39MB
(1M, 20k, 8) = ~0.6MB
(800, 8, 61400) = ~1.90MB
1. the benchmark script runs 100 trials for each group of indices, so if the cache's size is greater than the working set size, then the source data can be totally loaded in cache after paying for compulsory misses during the 1st trial, then after 1st trial you're actually measuring the cache bandwidth.
2. Although the benchmark runs 100 trials for the same input, users usually on run once for each input so they always pay for the compulsory misses and the bottleneck is the memory speed.
3. With a bit of search I found Skylake 8180 has a 38.5MB L3 cache, so both (800, 8, 61400) and (1M, 20k, 8) workloads can totally be cached into the L3 cache, so the measurements performed on those cannot accurately compare the performance on increasing memory bandwidth consumption of different versions of code.
4. If you really want to showcase the effect of different versions on num_cols=8, maybe you can switch to a CPU with smaller cache or you can increase test_rows to make the working set larger than your L3 cache size.
5. I tested (50M, 1M, 8) (~30.5M working set size, definitely greater than my L3 cache size) on my own machine and got 7.90 GB/s for "for" version and 9.80 GB/s for "memcpy" version.

**Comment by pengzhao-intel** [comment]
@HyperZealot I suggest to using an environment `MXNET_CPU_PARALLEL_COPY_SIZE` to decide when use `memcpy` or OMP copy as in #12926. The num_cols=8 case in here is just an indicator to cover different scenario.

**Comment by HyperZealot** [comment]
@pengzhao-intel I think what @rongzha1 is suggesting here is to have a special case for smaller sizes?
I've shown that "memcpy" could give more bandwidth than 'for' even when num_cols is small so maybe such special handling is indeed unnecessary? For your comment, are you suggesting that we change the kernel launch to omp loop when num_cols is large?

**Comment by HyperZealot** [comment]
@pengzhao-intel Thanks for your approval!

**Pull request merged**